### PR TITLE
Initial commit of Dapr Actor

### PR DESCRIPTION
**Pull request by youngbupark**
# Description

This is the initial commit of Dapr Actor for Python SDK. Initially, I have started this work to understand Virtual Actor implementation of dotnet-sdk Actor SDK 5 months ago. Therefore, the design are similar to dotnet-sdk port except for `ActorRuntimeContext` concept(leveraging from java-sdk) instead of `ActorService`.

* Typing: Python has been known as a duck-typing language. Since Python 3.5, it has supported [static type checking](https://docs.python.org/3/library/typing.html). I leveraged this feature to actor implementation to be consistent with dotnet-sdk usage as possible as I can; I believe that user will use the latest Python 3 instead of Python 2 which is deprecated.
* Current status
  - Implement initial skeleton of actor (ActorRuntime, ActorManager, ActorProxy, etc)
  - Support only actor method invocation feature
  - Support rpc-style method invocation (thanks to python's duck-typing)
  - Flask-integration example - service invocation
  - Added unit-tests even if most of tests are positive testing. :)
* Missing features
  - State Management
  - Reminder
  - Timer
  - Packaging

**Python Actor SDK is not supported officially by Dapr core team.** I hope more contributors are interested in python-sdk.

## Issue reference

n/a

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Extended the documentation

### CODE REVIEW

File: dapr/actor/runtime/actor.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+"""
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT License.
+"""
+
+from dapr.actor.actor_interface import ActorInterface
+from dapr.actor.runtime.methodcontext import ActorMethodContext
+from dapr.actor.runtime.context import ActorRuntimeContext
+
+class Actor:
```
Comment (artursouza) [inline dapr/actor/runtime/actor.py]: Use async/await everywhere: https://docs.python.org/3/library/asyncio.html

File: dapr/actor/id.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        This is a thread-safe operation that generates new random :class:`ActorId`
+
+        :returns: ActorId with random id
+        :rtype: :class:`ActorId`
+        """
+        random_id = ""
+        with cls._rand_id_lock:
+            random_id = secrets.token_hex(8)
+
+        return ActorId(random_id)
```
[the remark this answers is not in the document]
Comment (youngbupark) [reply dapr/actor/id.py]: Yeah. Used uuid here.

File: dapr/actor/runtime/actor.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        # TODO: Reset state
+        self._on_activate()
+
+        # TODO: Save state modification
+
+    def on_deactivate_internal(self):
+        # TODO: Reset state
+        self._on_deactivate()
+
+    def _on_activate(self):
```
[the remark this answers is not in the document]
Comment (youngbupark) [reply dapr/actor/runtime/actor.py]: _xxx is treated as a protected. so it make senses to have underscore prefix. won't fix.

File: dapr/actor/runtime/actor.py (hunk 1 above)
Comment (youngbupark) [reply dapr/actor/runtime/actor.py]: Yeah I switched to async methods. please review the code again.

File: dapr/actor/runtime/actor.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    TODO: Support Timer, Reminder, State Management
+
+    """
+
+    def __init__(self, ctx: ActorRuntimeContext, actor_id: str):
+        self.id = actor_id
+        self._runtime_ctx = ctx
+        self._dispatch_mapping = {}
+
+    def on_activate_internal(self):
```
[the remark this answers is not in the document]
Comment (youngbupark) [reply dapr/actor/runtime/actor.py]: Fixed.

File: dapr/actor/actor_interface.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+"""
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT License.
+"""
+
+from abc import ABC
+from dapr.actor.runtime.methodcontext import ActorMethodContext
+
+class ActorInterface(ABC):
```
[the remark this answers is not in the document]
Comment (youngbupark) [reply dapr/actor/actor_interface.py]: I would like to follow the original naming of Actor base class for now.

When it comes to removing IActor interface, I would like to keep dotnet-sdk design. To me, it makes sense when user derives the multiple interfaces and allow only methods in IActor-derived interfaces to be callable.

This python implementation also uses method decorator to rename the actor method name, but proxy can call any methods which does not have decorator if the methods is declared in ActorInterface-derived classes. e.g.
```
class DemoActor(ActorInterface)
def actor_method(self):
   ...
@actormethod(name="ActorMethod2")
def actor_method2(self):
  ...
```

python implementation can call actor_method and ActorMethod2 via proxy. actormethod is used only for renaming of method.

Won't fix.

cc/ @amanbha

**Comment by youngbupark** [comment]
@artursouza Please review this PR. I addressed all your comments and changed the target branch from master to features/actor.

**Comment by youngbupark** [comment]
@artursouza Thanks Artur! :)

**Pull request merged**